lib/core/text/u16_string.nit

```diff
@@ -0,0 +1,162 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Wrapper of ICU's UTF-16 encoded strings and conversion
+# This module is meant to ease the use of complex string operations provided by the ICU library.
+# The module provides a wrapper for ICU's string structure : `UChar *` as well as conversion functions to/from `String` and `CString`
+module u16_string is pkgconfig ("icu-io", "icu-i18n", "icu-uc")
+
+`{
+	#include <assert.h>
+	#include <unicode/utypes.h>
+	#include <unicode/ustring.h>
+
+	#include <unicode/uiter.h>
+`}
+
+# UTF-16 encoded string
+class U16String
+	super Finalizable
+	super Text
+
+	# Pointer to a `UChar *` string
+	private var uchar_string: nullable UCharString
+
+	# Number of code units `self` has available
+	private var capacity = 0
+
+	# Returns an empty `U16String` of length `len` or a NULL `U16String` if no `len` parameter is provided
+	init (len: nullable Int) do
+		if len == null then
+			uchar_string = new UCharString.nul
+		else
+			assert len >= 0
+			uchar_string = new UCharString.empty(len)
+			capacity = len - 1
+		end
+	end
+
+	# Returns a converted `U16String` from a `String`
+	init from_string(source: String) do
+		var csource = source.to_cstring
+		var csource_length = source.to_cstring.cstring_length
+
+		uchar_string = new UCharString.nul
+		var required_length = uchar_string.from_cstring(0, csource, csource_length)
+
+		uchar_string = new UCharString.empty(required_length + 1)
+		uchar_string.from_cstring(required_length + 1, csource, csource_length)
+		capacity = required_length
+	end
+
+	redef fun length: Int do return uchar_string.length
```

Comment:
Since `UCharString` does not know its own size,  `U16String` should keep the capacity of the buffer. And it does it seems since you have `capacity`.
So, when you ask a service to `uchar_string` you can pass the capacity as argument.